Finn/controllers

FinnMaladay opened the pull request:
Simple review to see if the direction (structure and application of code) is correct

FinnMaladay committed:
Index.html created in resources/views

index.html file used to demonstrate controller use to access database

FinnMaladay committed:
Commit small changes

FinnMaladay committed:
Merge branch 'main' into finn/Controllers

FinnMaladay committed:
Controller Structures Started

All blade pages created

index, show pages started; routes/controllers linked with blade php

FinnMaladay committed:
Show page created

Show route and page created. No style ye, simply shows data off specific movie selected

FinnMaladay committed:
Updated Controller

Controller updated:

Controller Functions

Create function has been made
Store function has been made
Destroy function has been made

Routes:

Route created to movies.create view
Route link created between movies.index and movies.create via button press (in movies.index)

Basic view created with movies.create
- Wanting to test function with 1 item first (movie title).
Add more after once initial testing is deemed successful

FinnMaladay committed:
Merge branch 'main' into finn/Controllers

vrodriguero commented on resources/views/movies/create.blade.php:
```diff
@@ -0,0 +1,49 @@
+
+        <!-- Styles -->
+        <style>
+            /*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */html{line-height:1.15;-webkit-text-size-adjust:100%}body{margin:0}a{background-color:transparent}[hidden]{display:none}html{font-family:system-ui,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;line-hei ...
```
I would suggest having a layout set for all your views, this not only makes things more clear but also reduces code duplication and improves performance. 

You installed the Laravel breeze and out of the box, there is an app layout for you to use it. So maybe start with that one, and customise if you need 😄

vrodriguero replied:
I believe that it would probably be best to have the movies done completely @FinnMaladay, before moving to the next todo. This is a good time for you to improve and clean up some of the code you think will be duplicated over and over again. 

- Layout is an example of it, your page does not need to have the final design. But the main layout of the website should be done, so you and Libro can create at least some consistency between pages. 
- Maybe get some of the HTTP tests done too before moving to the other controller. There are some examples of the tests in the TDD lesson and Laravel docs ;D. 
- That will make it much faster for you to implement the other features and relations.

FinnMaladay committed:
Error in Movie Controller

Error in movie Controller regarding Release_Date

"Carbon\Exceptions\InvalidFormatException
The separation symbol could not be found Trailing data"

vrodriguero committed:
save movie fix

FinnMaladay committed:
Basic Movie Controller completed

Movie controller can perform all CRUD actions

Index: Will display list all all movies currently avaliable -> need to update so it only display specific movies based on provided information, for example returning only movies that match a provide name (though does not need to be exact)

show: Will display a specific movie selected by user, with information

- Title
- Release date
- Poster
- Trailer (does not currently display, need to find better solution)
- Blurb

Still need to be able to display the genre and celebrities that are related to this movie

Edit/Update: When pressed, it will display all inform related to current movie and will allow user to change all information:

Need to create proper validate requirements and create validate file so admin cannot input invalid data.
Must also provide a way to check if the current user is an admin so that the edit/delete button may be displayed

Create: Will load a new page that present input containers for

Title
Release_date
Poster(url)
Trailer(url)
Blurb

Like Edit/Update, need to create proper validators to ensure invalid data isn't inserted

Delete: On movie page, if delete is selected, current movie is soft deleted from webpage (while remaining in the Database)

For all Post, Put and Delete actions, need to have a message display when the action has been performed successfully

vrodriguero commented on app/Http/Controllers/MovieController.php:
```diff
@@ -2,9 +2,74 @@
+            'release_date' => Carbon::parse(request()->input('release_date'))->format('d-m-Y'),
+            'poster' => request()->input('poster'),
+            'trailer' => request()->input('trailer'),
+            'blurb' => request()->input('trailer')
```
I believe this should be `'blurb'` 😄

FinnMaladay replied:
Changed to input('blurb')

vrodriguero commented on app/Http/Controllers/MovieController.php:
```diff
@@ -2,9 +2,74 @@
+            'poster' => ['required', 'string', 'min:16'],
+            'trailer' => ['required', 'string', 'min:16'],
+            'blurb' => ['required', 'string']
+        ]));
```
You can add a blank line before the return 😄 , this makes the code a bit cleaner

vrodriguero replied:
I believe for this one you still want to have the release date saving in the correct format. So for now, just do the same you have done in the `store()` method. 😸

FinnMaladay replied:
Changes made

vrodriguero commented on app/Http/Controllers/MovieController.php:
```diff
@@ -2,9 +2,74 @@
+    public function index() //print all data in Moives database
+    {
+        return view('movies.index', [
+            'movies' => Movie::all()
```
It is okay to return all, but we usually want to make sure that the data is presented in a useful manner. 
To identify these here is are some easy questions that will help you out:
- What information do I want to show on the index page?
- How should it be ordered?
- What about if we had a million movies? How would I show the records? Would it be paginated?

FinnMaladay replied:
Change the request

It now paginates 5 movies per page

It will also pass in the movies by order of name first (ascending order) and then release date (descending order)

vrodriguero commented on app/Http/Controllers/MovieController.php:
```diff
@@ -2,9 +2,74 @@
+            'trailer' => ['required', 'string', 'min:16'],
+            'blurb' => ['required', 'string']
+        ]));
+        return redirect()->route('movies.show', $movie)->with('success', 'Movie added');
```
I believe the message should be something related to the update action 😃

vrodriguero commented on routes/web.php:
```diff
@@ -17,8 +19,26 @@
     return view('dashboard');
 })->middleware(['auth'])->name('dashboard');
 
+
+
```
you can delete these empty lines ;D

vrodriguero reviewed (changes requested):
Hi @FinnMaladay! 

Good job!

Your effort on the controllers has paid off! 
I hope this is getting a bit more clearer as you move further in the course. 

I have made some suggestions in the comments ;D for you. But, I that before moving forward with the other relationships. I think it might be best for you to make some optimisations in the current code :D and start implementing some tests to make sure everything works as expected. 

This will make it much easier for you to scale your code and implement the remaining features.

FinnMaladay committed:
Initial changes to Movie Controller

Updated syntaxical errors and Movie::create() and $movie->update() function calls in controller

FinnMaladay committed:
Update to movie blade pages

Breeze layout and components applied to various movie blade pages

Notes

Breeze Guest Layout: Try and modify it so it includes the navigation bar

Index: App layout applied (need to figure out how to design so it loads either guest or user UI)

Show: Akin to index

Edit/Add: now both implement Breeze's input component for input data, both text and date
Need to work further on Edit's Release_date component: Value=old() isn't currently showing release date that exists in the database

FinnMaladay committed:
Pull Request updates

Updating errors in message outputs and code structure

vrodriguero commented on app/Http/Controllers/MovieController.php:
```diff
@@ -2,9 +2,73 @@
 
+use App\Models\Movie;
 use Illuminate\Http\Request;
 
```
You can remove this blank line @Finn 😄 
And also are we using the DB Facade anywhere? If not we can delete it.

vrodriguero commented on app/Http/Controllers/MovieController.php:
```diff
@@ -2,9 +2,73 @@
+        return view('movies.index', [
+            'movies' => Movie::orderBy('title', 'ASC')
+            ->orderBy('release_date', 'DESC')
+            ->paginate(5)
```
Maybe it will be better to have a bit more movies showing at a time?

Just think in scenarios where this might affect the user experience. If you are browsing to see what movies you have... and the DB has 50K movies, it will take some time for you to reach the end. 

I would same maybe change it to 25? or 50? You can always update the seeding 😄

FinnMaladay replied:
Seeding has been change: 1000 movies generated. Each page displays 25 at this stage. Might increase/decrease when visuals of page are updated a bit

vrodriguero commented on app/Http/Controllers/MovieController.php:
```diff
@@ -2,9 +2,73 @@
+    public function show(Movie $movie)
+    {
+        return view('movies.show', compact('movie'));
+        // similar to return view('movies.show', $movie=>movie);
```
I think we can remove this comment.

FinnMaladay replied:
comment removed

vrodriguero commented on resources/views/movies/create.blade.php:
```diff
@@ -0,0 +1,60 @@
+
```
Try to extend the existing layout here :D. This might give an error now, if you visit as a guest, about the $name as the App layout is using a component named navigation that gets the name from the authenticated user. 

However, I do think that for this route the user should be authenticated to be able to access this page don't you think? For this to work, you can use middleware. Laravel has a middleware called auth, which will only allow the user to visit a page if the user has been authenticated. Otherwise, it will redirect to the login page. 😄

FinnMaladay replied:
I have set the create and update pages to only have guest.layout as middle ware ensures that they require a log in still and neither of them have access to the navigation bar (there is no need for them to have access to it)

However, I might need to consider adding a 'Cancel' or 'Return' button if the admin changes their mind about the action

vrodriguero commented on resources/views/movies/edit.blade.php:
```diff
@@ -0,0 +1,31 @@
+
```
You can also use the layout here ;D

And we do not need a view to delete a movie we can just have a button on the index that only an admin can see or something.

vrodriguero commented on resources/views/movies/edit.blade.php:
```diff
@@ -0,0 +1,31 @@
+            <x-input type="text" name="trailer" id="trailer" :value="old('trailer', $movie->trailer)"/>
+            <div>
+            <!--Blurb-->
+            <x-input type="text" name="blurb" id="blurb" :value="old('blurb', $movie->blurb)"/>
```
Maybe for the blurb we can use a textarea? Seeing as this one is a longer text?

vrodriguero commented on resources/views/movies/index.blade.php:
```diff
@@ -0,0 +1,20 @@
+<x-app-layout>
```
As I mentioned before this might throw an error for unauthenticated users. So we might need to modify the navigation bar a bit to accommodate guests. Or alternatively, you can try to use the guest layout for this one?

FinnMaladay replied:
Mangage to get the naviagation bar to work for index and show, showing in the top right corner either login/register if the user is a guest or the user name if the user is currently logged in

vrodriguero commented on resources/views/movies/show.blade.php:
```diff
@@ -0,0 +1,41 @@
+<x-app-layout>
```
Same comment as the above 😄

vrodriguero commented on routes/web.php:
```diff
@@ -1,6 +1,8 @@
 <?php
 
+
```
You can delete this extra line here ;D

vrodriguero commented on routes/web.php:
```diff
@@ -17,6 +19,22 @@
 });
 
+//
+Route::get('movies', [MovieController::class, 'index'])->name('movies.index');
+
+Route::get('movies/create', [MovieController::class, 'create'])->name('movies.create');
+
+Route::get('movies/{movie}/edit', [MovieController::class, 'edit'])->name('movies.edit');
+
+Route::get('movies/{movie}', [MovieController::class, 'show'])->name('movies.show');
+
+Route::post('movies', [MovieController::class, 'store'])->name('movies.store');
+
+Route::put('movies/{movie}', [MovieController::class, 'update'])->name('movies.update');
+
+Route::delete('movies/{movie}', [MovieController::class, 'destroy'])->name('movies.destroy');
```
As I mentioned in one of the comments above, I think that we should check which of these routes can be accessed by a user and which can be accessed by a guest. Take a look at the Laravel docs and check for `Route::group`, you can group your routes and use the same middleware. In this case, you can separate them in guests and auth.

FinnMaladay replied:
I could not find the Route::group that you made mention of. The closest that I could find in the laravel documentation is Route::middleware. But doing this still manage to seperate some of the routes between required auth routes and regular 'guest' routes

vrodriguero replied:
https://laravel.com/docs/8.x/routing Check the docs :D they have some examples of a route group there

vrodriguero replied:
Yes, that is correct @FinnMaladay, to use a middleware in a group you do need to use the `Route::middleware()` :D you are looking in the right place. And the example they have can be applicable in your scenario too. 

The `Route::group` is used in that example as well. Just look a bit closer and you will see that group() is chained right after the middleware().

vrodriguero reviewed (changes requested):
Hi @FinnMaladay Great effort!

I left some comments and suggestions, but don't hesitate to send me a ping if you did not understand any of them. 

😄  Happy coding!

FinnMaladay committed:
Movie entity updates

Form component created - Will be used for all future uses of <form>

User Factory implemented in Seeder - password now has default value that is encrypted

Some changes made to breeze navigation bar - further changes need to be made so it displays user or login text

vrodriguero committed:
- Use guest layout for index and show
- Fix missing close div on index
- Change video tag to iframe
- Updated movie factory to have an embed url instead of the common one.

FinnMaladay committed:
Minor updates

edit now uses breeze login design and structure. Further editing need

remove un-need code

FinnMaladay committed:
Fixed CRUD errors

Had some errors with crud function (specifically Create). Now all functions work correctly.

Will move onto trying to implement middleware for the CRUD so only authenticated users can access them

FinnMaladay committed:
Delete modal started

> Delete:

Delete view blade has been created: Created to set basic design of modal.
Need to set up blade into component and implement into movies.show() using Apline as well

Source of modal code: https://www.mywebtuts.com/blog/alpine-js-tailwind-modal-example

FinnMaladay committed:
Navigation bar implement

Navigation bar is now implemented

It will display on all expect pages

for example

Dashboard

Moive.Index

Movie.Show

For guest, it will show 'login' and 'register' in the top right corner of the screen. Once they are logged in, these elements will be replaced by the users full name, with the option of a drop menu, currently with a single item of 'log out'

Further changes are still needing to be made to the navigation bar; namely, when the user logins in from movie.index or movie.show, it will redirect them to dashboard page. Need to change it so it leaves the user on the current page that they were in

FinnMaladay committed:
Features tests for Movies

Series of simple feature tests created for Movies controller functions

o MovieShowTest

2 simple tests test..:

- Screen can render: Using an instance of $movie, containing test data based on Movie Database..
Will check that the route for (movie.show) will load for $movie->id

- 404 non-existent movie page: Will test for failure to load movie page when an invalid $movie->id is provided

o MovieIndexTest

1 simple test

- Screen can render: Check that index page can be asserted (returning HTTP 200 code)

o MovieCreateTest

4 tests for create function

- Render with authorised user: Checks that page will load when authorised uses selects it

- new movie added: Checks when valid date is provide (meeting the controller validation) no errors are thrown and when data is created, will properly redirect

2 fail test created

- Invalid data: Data past to poster does not match the validation requirements. Error will be thrown

- No data provided: 'title' data is not provide. Error will be thrown

o MovieEditTest

2 access based test

- Screen can be rendered: When a valid user is signed in, the edit page may be accessed and rendered, returning 200 code

- Page cannot be accessed: Test to fail, edit page attempted to be access when user is not signed throws error

o MovieDeleteTest

1 simple test

- Authorised user can delete

Check that only a valid logged in user may delete a $movie entity, then properly redirected to correct URI

vrodriguero commented on resources/views/layouts/navigation.blade.php:
```diff
@@ -12,14 +12,60 @@
 
                 <!-- Navigation Links -->
                 <div class="hidden space-x-8 sm:-my-px sm:ml-10 sm:flex">
+
+                    <!-- Update this so route goes to the result/search screen -->
```
I did not quite understand this. Why are we adding substituting the search page? 

This one is also redirecting to the dashboard, which does not make much sense. 

Probably better to keep the dashboard and add an additional tab for the search when we implement that.

FinnMaladay replied:
Will consider adding an addition tab later for search bar implementation 

For now, changed tag to 'Movies' and now routes to 'movies.index' for the time being

vrodriguero commented on resources/views/layouts/app.blade.php:
```diff
@@ -21,11 +21,12 @@
             <!-- Page Heading -->
-            <header class="bg-white shadow">
+                {{-- original --}}
+            {{-- <header class="bg-white shadow">
                 <div class="max-w-7xl mx-auto py-6 px-4 sm:px-6 lg:px-8">
                     {{ $header }}
                 </div>
-            </header>
+            </header> --}}
```
This was commented out to match the wireframe? ;D 

Just check if there are any other pages that are using this component and that might be passing the header slot.

FinnMaladay replied:
It was originally throwing an error inside movies.index which implemented app-layout. Have manage to fix it now (not sure exactly was throwing the error initially)

vrodriguero commented on resources/views/layouts/navigation.blade.php:
```diff
@@ -12,14 +12,60 @@
+                </div>
+            @endif
+
+                {{-- original --}}
+            {{-- if the user is logged in --}}
+            {{-- <div class="hidden sm:flex sm:items-center sm:ml-6">
```
Feel free to remove any lines that you are not using ;D we can always compare the file with the previous version, in case we need to revert the change to a previous commit.

FinnMaladay replied:
Okay, I will. Just a force of habit to no accidentally delete code incase everything goes horrible wrong

vrodriguero commented on resources/views/layouts/navigation.blade.php:
```diff
@@ -69,7 +115,34 @@
 
         <!-- Responsive Settings Options -->
-        <div class="pt-4 pb-1 border-t border-gray-200">
+        @if (Route::has('login'))
+            <div class="hidden fixed top-0 right-0 px-6 py-4 sm:block">
+                @auth
+                <div class="pt-4 pb-1 border-t border-gray-200">
+                    <div class="px-4">
+                        <div class="font-medium text-base text-gray-800">{{ Auth::user()->name }}</div>
+                        <div class="font-medium text-sm text-gray-500">{{ Auth::user()->email }}</div>
+                    </div>
+
+                    <div class="mt-3 space-y-1">
+                        <!-- Authentication -->
+                        <form method="POST" action="{{ route('logout') }}">
+                            @csrf
+
+                            <x-responsive-nav-link :href="route('logout')"
+                                    onclick="event.preventDefault();
+                                                this.closest('form').submit();">
+                                {{ __('Log Out') }}
+                            </x-responsive-nav-link>
+                        </form>
+                    </div>
+                </div>
+                @endauth
+            </div>
+        @endif
```
we could probably have a similar approach to the one in the above ;D where if the user is a guest we can have the login, register routes there

FinnMaladay replied:
I have have add the same lines of code as above where if the user isn't logged in it presents the options of login or register with associated routes

vrodriguero commented on resources/views/movies/create.blade.php:
```diff
@@ -0,0 +1,58 @@
+            {{ __('Create a new film entity') }}
+        </h2>
+    </x-slot>
+    <x-auth-card>
```
I would probably try to create something similar to your wireframe, instead of having the authentication card component here. 

The `x-slot` header will not show and the reason is because of the changes on the x-app-layout.

FinnMaladay replied:
Just using the auth-card as a temporary layout. I plan on using something similar to it for the movie create and edit

vrodriguero replied:
Got it ;D
I would suggest you start looking at this now ;D otherwise you would have to change all the places you will use it, and sometimes that is hard to keep track :D

FinnMaladay replied:
Update design has been implemented. Might need some further improvements however add and edit functions all work accordingly and some minor responsive implementation for when the screen size gets smaller

vrodriguero commented on resources/views/movies/delete.blade.php:
```diff
@@ -0,0 +1,42 @@
+<body>
+
+    {{-- Need to convert this section of code into proper component --}}
+    <!-- modal div -->
+    <div class="mt-6" x-data="{ open: false }">
+        <!-- Button (blue), duh! -->
+        <button class="px-4 py-2 rounded select-none outline-black focus:shadow-outline" @click="open = true">DELETE</button>
+        <!-- Dialog (full screen) -->
+        <div class="absolute top-0 left-0 flex items-center justify-center w-full h-full" style="background-color: rgba(0,0,0,.5);" x-show="open"  >
+            <!-- A basic modal: Warning message with Cancel or Confirm button -->
+            <div class="h-auto p-4 mx-2 text-left bg-white rounded shadow-xl md:max-w-xl md:p-6 lg:p-8 md:mx-0" @click.away="open = false">
+                <!-- dialog box -->
+                <div class="mt-2">
+                    <p class="text-sm leading-5 text-gray-500">
+                        Do you wish to delete the following entity?
+                    </p>
+                </div>
+                <!-- Confirmation button -->
+                <div class="mt-5 sm:mt-6">
+                    <span class="flex w-full rounded-md shadow-sm">
+
+                        {{-- {{$slot}} Delete From will be inserted here--}}
+
+                        <button class="inline-flex justify-center w-full px-4 py-2 text-white bg-purple-500 rounded hover:bg-purple-700">
+                            CONFIRM
+                        </button>
+                    </span>
+                </div>
+                <!-- Cancel button --->
+                <div class="mt-5 sm:mt-6">
+                    <span class="flex w-full rounded-md shadow-sm">
+                        <button @click="open = false" class="inline-flex justify-center w-full px-4 py-2 text-white bg-red-500 rounded hover:bg-red-700">
+                            CANCEL
+                        </button>
+                    </span>
+                </div>
+            </div>
+        </div>
+    </div>
+</body>
```
You can also make modal a component and then call that on the page you use it and change the open and close. 

At the moment, this is not showing because you are posting to the delete route so it will delete directly, we want to wire that to open the modal. ;D

FinnMaladay replied:
The delete modal is something I am still working on. I simple wanted to start with a view page so i could see the design and see if It looked how I wanted it too

vrodriguero replied:
Okay ;D
How did you test to see if this looked how you wanted? 
I think that even if this is a view you would still need to use one of the layouts ;D

vrodriguero commented on routes/web.php:
```diff
@@ -15,8 +16,35 @@
+//     return view('home');
+// })->name('homepage');
+
 Route::get('/dashboard', function () {
     return view('dashboard');
 })->middleware(['auth'])->name('dashboard');
```
This can also be moved into the group 😄

FinnMaladay replied:
'/dashboard' Route call has been moved into group

vrodriguero commented on routes/web.php:
```diff
@@ -15,8 +16,35 @@
     return view('welcome');
+})->name('welcome');
+
+Route::middleware(['auth'])->group(function () {
```
Awesome!

vrodriguero commented on tests/Feature/AuthenticationTest.php:
```diff
@@ -24,14 +24,14 @@ public function test_users_can_authenticate_using_the_login_screen()
     }
 
-    public function test_users_can_not_authenticate_with_invalid_password()
+    public function test_users_can_not_authenticate_with_invalid_password() // Test fails with current test build
```
This seems to be working on my end ;D

FinnMaladay replied:
It was initially throwing a fail when it should have passed, but I discovered it was because the test was looking for the wrong password (initially "password" but it was changed to "123456")

vrodriguero commented on tests/Feature/MovieCreateTest.php:
```diff
@@ -0,0 +1,72 @@
+            'trailer' => 'https://www.youtube.com/watch?v=-1dSY6ZuXEY',
+            'blurb' => 'This is a sample blurb for the movie A Test movie'
+        ])->assertSessionHasNoErrors()
+        ->assertRedirect();
```
When testing if the new movie is created, we need to also check if the database has that record :D 

You can use the $this->assertDatabaseHas()....

FinnMaladay replied:
a test has been added implementing assertDatabaseHas(). It only checks for one item (that being one of the possible trailers) that exists. Need to check if I can check for all columns

i.e

`$this->assertDatabaseHas('movies', [
            'title' => ....
            'release_date'=> ....
            'poster'=>.......
            'trailer' => 'https://www.youtube.com/embed/LIfcaZ4pC-4'
             etc
        ]);`

vrodriguero commented on tests/Feature/MovieCreateTest.php:
```diff
@@ -0,0 +1,72 @@
+        ->assertRedirect();
+    }
+
+    public function test_movie_create_invalid_data_throws_error()
```
Looks okay :D Just one suggestion on the naming of the test. 

This should literally discribe what you are testing, so if you are testing  that the app throws validation error when you have invalid poster value. Your test should be named something like test_invalid_poster_movie_attribute_throws_validation_error.

FinnMaladay replied:
Name for various tests have been improved to more concisely and literally describe test

vrodriguero commented on tests/Feature/MovieCreateTest.php:
```diff
@@ -0,0 +1,72 @@
+        ->assertRedirect();
+    }
+
+    public function test_movie_create_invalid_no_data_provided_for_key_variables(){
```
Perhaps we could enhance the name for this test 😄

FinnMaladay replied:
Agreed as well, I have shorten it and more literally described the test

vrodriguero commented on tests/Feature/MovieDeleteTest.php:
```diff
@@ -0,0 +1,27 @@
+     *
+     * @return void
+     */
+    public function test_movie_delete_authurised_user_can_remove_movie()
```
Another test: 
- guest cannot delete movie
- movies that do not exist are not deleted.

Any tests that involve changing something in the database, we want to check if the database is the way we expected. :D

vrodriguero commented on tests/Feature/MovieShowTest.php:
```diff
@@ -0,0 +1,31 @@
+    {
+        $movie = Movie::factory()->make();
+
+        $response = $this->get('/movies/'.$movie->id);
+
+        $response->assertStatus(200);
```
you can just chain this one ;D and use assertOK() instead of the status one ;D

vrodriguero commented on tests/Feature/MovieIndexTest.php:
```diff
@@ -0,0 +1,22 @@
+     */
+    public function test_movie_index_screen_can_render()
+    {
+        $response = $this->get('/movies');
+
+        $response->assertStatus(200);
+    }
```
You can just chain this one ;D and use assertOk() instead of the assertStatus(200). Just for consistency

vrodriguero commented on tests/Feature/MovieEditTest.php:
```diff
@@ -0,0 +1,36 @@
+
+        $this->actingAs($user)
+        ->get('/movies/'.$movie->id.'/edit')
+        ->assertStatus(200);
```
assertOk() :D

vrodriguero commented on tests/Feature/MovieEditTest.php:
```diff
@@ -0,0 +1,36 @@
+        $movie = Movie::factory()->make();
+
+        $this->get('/movies/'.$movie->id.'/edit')
+        ->assertStatus(404);
```
This one I believe it should redirect to login no? The movie exists so it does not make sense that it is returning a 404. 
In fact, asserting that the user cannot edit an unexisting movie can be a test case as well.

FinnMaladay replied:
No exactly sure how I should conduct test this;

The test I wanted to do was a fail; where if the user isn't logged in it should no allow access to edit

My assumption was since a user needs to be logged in to access edit, any test that doesn't have `actingAs() ` in it would fail

Though it probably would be better to test that it redirects since that is how the program functions

Again though I am no exactly sure how to test this

FinnMaladay replied:
Thanks for your clarification on slack. Manage to get the test functional

vrodriguero commented on tests/Feature/MovieCreateTest.php:
```diff
@@ -0,0 +1,72 @@
+
+        $this->actingAs($user)
+        ->get('/movies')
+        ->assertStatus(200);
```
assertOk() can be used instead of assertStatus(200). It just makes things more readable.

vrodriguero reviewed (changes requested):
Hi Finn! 

Great effort so far!

 I have left some suggestions for you. 

Keep up the good work!

FinnMaladay committed:
Pull Request changes

Nearly all queries asked for pull request has been made.

All changes request for the tests have been made:

- Improving various test names to be more concise and literal

- Changing to assertOk() to increase readability

- assertRedirect() calls have been change to declare exact redirect (for clarity)

- Additional tests have been made in MovieDeleteTest.php

- Grouping in web.php has been improved

- Formatting changes across all blade pages for movie have been made (still a work in progress and needs to be improved further)

FinnMaladay committed:
Display Many to Many relations

Website will now display the many to many relationships of Celebrity_Movie and Genre_Movie

Movie.Show() will now display Celebrities and Genres that have a relationship with Movie/{id}.

Clicking on a celebrity will route the user to Celebrity.show().
However I do not current have access to the Celebrity routes or Celebrity pages, so a simple test Route::get has been created with test.blade. This should be deleted later

Clicking on a genre performs a similar action: Routing the user to Genre.Show()

Genre.Show() will show all movies that are of that genre type. Clicking on a movie will take the user back to Movie.show().

Additional functions still required for Genre:

Create() -> Create a relation between current genre and a Movie

Delete() -> Remove relation between current genre and movie

Drop down menu: From Genre.Show(), a drop down menu should exist that allows the user to change the Genre and movie relations on display

vrodriguero commented on app/Http/Controllers/MovieController.php:
```diff
@@ -2,9 +2,72 @@
+    public function show(Movie $movie)
+    {
+
+        $movie->load('celebrities');
```
You can do a shortcut and have that in one line ;D 
```
$movies->load(['celebrities', 'genres']);
```

FinnMaladay replied:
Done

vrodriguero commented on app/Http/Controllers/MovieController.php:
```diff
@@ -2,9 +2,72 @@
+    {
+        return view('movies.create');
+    }
+    public function store(Request $request)
```
Maybe add a blank line between functions to improve readability.

vrodriguero reviewed (commented):
Hi @FinnMaladay ! 

Keep up with the good work :D I have left some suggestions and responded to some of your comments ;D

FinnMaladay committed:
Display of movie Profile complete

HTML display of movie profile page completed (with Tailwind CSS)

Might need to re-evaluate  how the displaying the Celebrities and Genres work  (currently use padding that will reduce when the screen becomes smaller)

Need to further define flex for certain <div> components to ensure website looks correct when on a different screen size

FinnMaladay committed:
Add Movie layout updated and New test added

The layout of add movie (movie.create) has been updated to an original design. Some further improving is needed

New tests have been created/updated for Delete and Edit

Delete now test that an entity has been deleted properly with assertSoftDeleted()

Edit contains 3 new test

1. Asserting that (with valid data) an edit within the database has been properly updated (it can be found with assertDatabaseHas)

2. Testing that a movie entity that does not exist can not be editted

3. Testing that when invalid date is passed, an error is throw and the put call fails

vrodriguero commented on resources/views/movies/create.blade.php:
```diff
@@ -1,58 +1,93 @@
-        <h2 class="font-semibold text-xl text-gray-800 leading-tight">
-            {{ __('Create a new film entity') }}
+        <h2 class="font-semibold text-xl leading-tight">
+            {{ __('Add a new film to Database') }}
```
We usually avoid using language that only we would understand. :D So for example for a user that is not familiar with this concept it might be a bit confusing 😄 . 

You can rename this to `Add a movie` or `Create a movie`.

FinnMaladay replied:
Makes sense. I will update them accordingly

vrodriguero commented on resources/views/movies/create.blade.php:
```diff
@@ -1,58 +1,93 @@
+                            <x-label for="celebrity" value="Celebrities"/>
 
-                        <textarea name="blurb" id="blurb"></textarea>
+                            <textarea class="leading-normal border-pink-500 border bg-purple-800 bg-opacity-25"
```
This should probably be a `checkbox` instead of a `textarea`. The celebrities will need to be associated with a movie, the value of the checkbox would be a celebrity_id.

FinnMaladay replied:
Ah okay, I wasn't sure what the best way would have been to associating the celebrity with a movie given the size that the data potentially could have been

vrodriguero commented on resources/views/movies/create.blade.php:
```diff
@@ -1,58 +1,93 @@
+                            <x-label for="genre" value="Genres"/>
 
-                </div>
+                            <textarea class="leading-normal border-pink-500 border bg-purple-800 bg-opacity-25"
```
Same as the comment above ;D. 

How I would do it, I would get all the genres in the controller and then have a for each loop here to show the genres list in the form. Inside the loop, I would associate the value of it with the id of the genre. In the request, I would pass an array of genres_ids and I would validate them. If the validation passes, I would use the `syncWithoutDetaching()` method from the [Laravel docs](https://laravel.com/docs/8.x/eloquent-relationships#syncing-associations) to link the movie with the genres id I selected.

vrodriguero reviewed (commented):
Hi Finn! I have left some suggestions for you in the comments bellow 😄

FinnMaladay committed:
Add and edit pages updated

Custom layout has been created for Add and Edit page.

Functions are all identical as before

Minor responsive implemented: Reduces columns of contents from 2 to 1 so all input text box appear in a single file

Added in Celebrity and Genre TextArea for future Many to Many function implementation. Unsure if this is the best approach to take (how to ensure that movie or genre provide by Admin exists in database? Does a custom requirement need to be defined?)

FinnMaladay committed:
Push -14/10/2021

FinnMaladay committed:
Merge branch 'main' into finn/Controllers

FinnMaladay committed:
Edit function completed

Movie can now be fully created and updated including with many to many relationship tables

For genre and celebrity

FinnMaladay committed:
Website style updated

Website style has been updated

Update has been implemented into:
- app-layout
- nav-layout (both of these are universal to the theme/colour tone of the website)

- Movie.index
-Move.show
-Movie.create
-Movie.edit

vrodriguero committed:
- Remove old Laravel homepage route

vrodriguero committed:
- Update factory for shorter titles and update design

vrodriguero committed:
- used array helper for the random movie and poster

vrodriguero committed:
- removed fixed from user dropdown

vrodriguero committed:
- show success message
